### PR TITLE
Make EquatorialCompression generic in polar axis.

### PR DESCRIPTION
**Pull request by mar-celine**
## Proposed changes

The existing EquatorialCompression coordinate map compresses angles toward/away from the z axis.
The existing HalfWedges are aligned along the y-z plane - their axes of symmetry are different.

This PR addresses this incompatibility between HalfWedges and EquatorialCompression by allowing
the user to select which axes to treat as the polar axis.

This PR addresses some of issue #3545.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Domain/CoordinateMaps/EquatorialCompression.cpp [hunk 1]
```diff
@@ -54,42 +62,49 @@ EquatorialCompression::angular_distortion(const std::array<T, 3>& coords,
...
-  get<1, 2>(jacobian_matrix) = y * z * lambda2;
-  get<2, 2>(jacobian_matrix) = inverse_alpha * (z * z * lambda2 + lambda3);
+  const ReturnType zero = make_with_value<ReturnType>(x, 0.0);
+  for (size_t i = 0; i < 3; i++) {
+    for (size_t j = 0; j < 3; j++) {
+      const size_t k = (index_pole_axis + j) % 3;
+      jacobian_matrix.get(i, k) =
+          (i == index_pole_axis ? inverse_alpha : 1.0) *
+          (coords[i] * coords[k] * (j == 0 ? lambda2 : lambda1) +
+           (i == k ? lambda3 : make_with_value<ReturnType>(x, 0.0)));
```
Comment (nilsvu) [inline src/Domain/CoordinateMaps/EquatorialCompression.cpp]: If I understand this correctly, then I think it's better to split this addition into a separate statement under an `if (i==k)` clause, so you don't need the zero vector.

File: tests/Unit/Domain/CoordinateMaps/Test_EquatorialCompression.cpp [hunk 2]
```diff
@@ -46,36 +46,44 @@ void test_radius() {
...
-  const CoordinateMaps::EquatorialCompression angular_compression_map(
-      aspect_ratio);
-  const auto result_point = angular_compression_map(input_point);
-  const auto& result_x = result_point[0];
-  const auto& result_y = result_point[1];
-  const auto& result_z = result_point[2];
 
-  // The EquatorialCompression map should preserve radial lengths.
-  CHECK(magnitude(result_point) == approx(magnitude(input_point)));
+  for (const auto& index_polar_axis : {0ul, 1ul, 2ul}) {
```
Comment (nilsvu) [inline tests/Unit/Domain/CoordinateMaps/Test_EquatorialCompression.cpp]: (optional) You can use `1_st`, `2_st` etc for size_t literals, as defined in Utilities/Literals.hpp.

File: src/Domain/CoordinateMaps/EquatorialCompression.hpp [hunk 3]
```diff
@@ -26,20 +26,24 @@ namespace CoordinateMaps {
...
- * at an angle \f$\theta\f$ from the pole are mapped to a new angle
+ * \details A mapping from the sphere to itself which redistributes points
+ * towards a user-specifed axis, indicated by `index_pole_axis_`.
+ * Once the axis is selected, the map is determined by a single parameter,
+ * the `aspect_ratio` \f$\alpha\f$, which is the ratio of the distance
+ * perpendicular to the polar axis to the distance along the polar axis for
+ * a given point. This parameter name was chosen because points with
+ * \f$\tan \theta = 1\f$ get mapped to points with \f$\tan \theta' = \alpha\f$.
+ * In general, gridpoints located at an angle \f$\theta\f$ from the pole are
+ * mapped to a new angle
```
Comment (nilsvu) [inline src/Domain/CoordinateMaps/EquatorialCompression.hpp:37]: Question: Do I understand correctly that this map only allowed the compression in one angular dimension of a sphere before, but now you want to apply it to both angular dimensions? So, will you apply this map twice? How is this change related to the half-wedges that you mention in the PR description?

**Review by nilsvu — CHANGES_REQUESTED** [review]

File: src/Domain/CoordinateMaps/EquatorialCompression.hpp (hunk 3 above)
Comment (mar-celine) [reply src/Domain/CoordinateMaps/EquatorialCompression.hpp:37]: The angle theta that is referenced here is the pole angle (the z axis in the current code), the azimuthal angle of mapped coordinates is left unchanged.

This PR allows the pole axis to be changed to a different axis, and the pole angle of a mapped coordinate will change with respect to that axis (the azimuthal angle of a mapped coordinate with respect to that axis is left unchanged).

Another way to think about the mapping being done is that if the existing map is EC(x), the new map is R^{-1}EC(R(x)), where R is a rotation matrix that rotates the z axis to the user specified axis.

The Half Wedges in spherical domains, including the BinaryCompactObject domain, are obtained by splitting ordinary Wedges in half down the y-z plane, which leads to a compressed grid close to the y-z plane. This can be undone by applying an EC map with the pole axis set to be the x axis, so that the equator is the y-z plane. The current EC map's pole axis is hard coded to be the z axis. With this addition the two options of Half Wedges and EC can be used together to obtain a spherical shell of 10 (Half) Wedge blocks where each (Half) Wedge has the same angular size on the sphere.

File: src/Domain/CoordinateMaps/EquatorialCompression.cpp (hunk 1 above)
Comment (mar-celine) [reply src/Domain/CoordinateMaps/EquatorialCompression.cpp]: This addition is inside the multiplication by `(i == index_pole_axis ? inverse_alpha : 1.0)` so I will also need to factor that out of this loop as well, so it would look like:

```
jacobian_matrix.get(i, k) =
          coords[i] * coords[k] * (j == 0 ? lambda2 : lambda1);
if(i== k){
    jacobian_matrix.get(i, k) += lambda3;
}
if(i == index_pole_axis){
    jacobian_matrix.get(i, k) *= inverse_alpha;
}
```

I can change it to this, it also looks clearer

File: src/Domain/CoordinateMaps/EquatorialCompression.cpp [hunk 4]
```diff
@@ -54,55 +62,65 @@ EquatorialCompression::angular_distortion(const std::array<T, 3>& coords,
...
-  // Various common factors:
-  const ReturnType lambda1 = (pow<2>(inverse_alpha) - 1.0) * z * z;
-  const ReturnType lambda2 = (1.0 - pow<2>(inverse_alpha)) * (x * x + y * y);
-  const ReturnType lambda3 = pow<2>(rho * radius);
-  const ReturnType lambda4 = radius * pow<3>(rho);
+      sqrt(square((index_pole_axis == 0) ? inverse_alpha * x : x) +
+           square((index_pole_axis == 1) ? inverse_alpha * y : y) +
+           square((index_pole_axis == 2) ? inverse_alpha * z : z));
 
+  // Various common factors:
```
Comment (nilsvu) [inline src/Domain/CoordinateMaps/EquatorialCompression.cpp]: Delete this line while squashing

File: src/Domain/CoordinateMaps/EquatorialCompression.hpp (hunk 3 above)
Comment (nilsvu) [reply src/Domain/CoordinateMaps/EquatorialCompression.hpp:37]: Ok I think I understand now, thanks for explaining! We take the x=0 plane as the equatorial plane and re-distribute grid points _away_ from it, to make the half-wedges larger.

What I don't yet understand: When the original map was equiangular, then the equatorially compressed (or elongated) map will not be equiangular anymore, or will it? Is it feasible to restore the equiangular property? Intuitively it seems to me that we don't want to continuously distort the angular dimension so the density of points near the equator is larger than near the poles (which I think the equatorial elongation does). Instead, I think we would like to keep the density of points constant along the polar angle ("equiangular") and divide up the polar dimension into equally-sized blocks. Specifically, split the outer shell at polar angle 30deg, at 90deg (the equator, i.e. the x=0 plane), and at 150deg. For the `Wedge` map this seems possible by specifying the opening angle in both angular dimensions. For `Frustum` I don't yet understand the map well enough to know if it's feasible to begin a radial grid line at a corner of the inner cube and end it at polar angle 30 / 90 / 150, instead of the current 45 / 90 / 135. Does what I'm thinking out loud here make sense at all? Is it perhaps already solved by the equatorial compression map, or can it, or do we need to build this into `Wedge` and `Frustum`? Please let me know what you think.

**Review by nilsvu — DISMISSED** [review]

File: src/Domain/CoordinateMaps/EquatorialCompression.hpp (hunk 3 above)
Comment (mar-celine) [reply src/Domain/CoordinateMaps/EquatorialCompression.hpp:37]: The equiangular vs equidistant distinction can be boiled down to the computation domain(CD) metric term `g_{\xi \xi}` to see how changes in the logical coordinate relate to distances in the physical domain. Wanting this term to be constant (indicating logical distances map linearly to physical distances along a (curved) surface) determines whether equidistant or equiangular coordinates should be used. Check out [redistributing gridpoints](https://spectre-code.org/redistributing_gridpoints.html) to see how the equiangular distribution can be solved for.

Using an Equatorial Compression (EC) map on a map this is equiangular (EA) to begin with (constant computational domain metric along the spherical surface) will lead to a non-constant CD metric. But whether it is still a "good" or "better than equidistant" map is determined by the behavior of the CD metric, perhaps not all non-constant CD metrics are "just as bad" as using an equidistant(ED) map. (I still need to look into this, but my intuition is that an EC+EA map is still better than an ED map, even if EC+EA != EA.)

I think an EC+EA map might lead to a distribution of points which is "locally EA" which would still be better than ED.
I think there is a value of `aspect_ratio` for the EC such that the Half Wedges have the same angular size as the full Wedges - such that a Shell of 2 Wedges + 8 Half Wedges will have 10 Blocks all of the same angular size on the sphere, with a gridpoint distribution that reflects that. If I understand correctly I think this is (close to) what you want? I am still thinking about your second paragraph but let me know if what I'm saying helps you narrow down what we might want for the domain!

**Comment by nilsvu** [comment]
@gsb76 Could you rebase this on current develop please?

File: src/Domain/CoordinateMaps/EquatorialCompression.cpp [hunk 5]
```diff
@@ -53,55 +62,64 @@ EquatorialCompression::angular_distortion(const std::array<T, 3>& coords,
...
-  get<2, 1>(jacobian_matrix) = inverse_alpha * y * z * lambda1;
-
-  get<0, 2>(jacobian_matrix) = x * z * lambda2;
-  get<1, 2>(jacobian_matrix) = y * z * lambda2;
-  get<2, 2>(jacobian_matrix) = inverse_alpha * (z * z * lambda2 + lambda3);
+  for (size_t i = 0; i < 3; i++) {
+    for (size_t j = 0; j < 3; j++) {
+      const size_t k = (index_pole_axis + j) % 3;
+      jacobian_matrix.get(i, k) =
+          coords[i] * coords[k] * (pow<2>(inverse_alpha) - 1.0);
```
Comment (knelli2) [inline src/Domain/CoordinateMaps/EquatorialCompression.cpp]: When accessing `std::array`s inside loops, use `gsl::at(coords, i)`. Same everywhere else.

File: src/Domain/CoordinateMaps/EquatorialCompression.hpp [hunk 6]
```diff
@@ -91,17 +97,20 @@ class EquatorialCompression {
  private:
   template <typename T>
   std::array<tt::remove_cvref_wrap_t<T>, 3> angular_distortion(
-      const std::array<T, 3>& coords, double inverse_alpha) const;
+      const std::array<T, 3>& coords, double inverse_alpha,
+      size_t index_pole_axis) const;
```
Comment (knelli2) [inline src/Domain/CoordinateMaps/EquatorialCompression.hpp]: Why do `angular_distortion` and `angular_distortion_jacobian` below take the `inverse_alpha` and `index_pole_axis` arguments? The functions are only used internally and the values passed to `inverse_alpha` and `index_pole_axis` are the member variables `aspect_ratio_` and `index_pole_axis_`, respectively.  Can't you just use the member variables within the functions rather than passing them?

File: tests/Unit/Domain/CoordinateMaps/Test_EquatorialCompression.cpp [hunk 7]
```diff
@@ -46,36 +47,44 @@ void test_radius() {
...
+    // The EquatorialCompression map should preserve radial lengths.
+    CHECK(magnitude(result_point) == approx(magnitude(input_point)));
+
+    // It is easy to mix up \alpha with one over \alpha,
+    // so this comment serves as a reminder to future readers
+    // as to which direction the \alpha should go in the
+    // CHECK statement.
+    //
+    // The aspect ratio is the ratio of the horizontal length
+    // to the vertical height. The parameter name `aspect_ratio` is
```
Comment (knelli2) [inline tests/Unit/Domain/CoordinateMaps/Test_EquatorialCompression.cpp]: I know you didn't change any wording here, but could you clarify "horizontal length" and "vertical height" here? Maybe using terminology similar to what you used in the doxygen description of `EquitorialCompression.hpp`

**Review by knelli2 — CHANGES_REQUESTED** [review]

File: src/Domain/CoordinateMaps/EquatorialCompression.hpp (hunk 6 above)
Comment (mar-celine) [reply src/Domain/CoordinateMaps/EquatorialCompression.hpp]: the coordinate map and the inverse coordinate map are the same, except for the value of `alpha` (the `alpha` for the inverse map is the inverse of the `alpha` for the forward map) so (inverse) `alpha` is left as an argument to `angular_distortion` (and `angular_distortion_jacobian`), which the (inverse) coordinate map and jacobian then call, where they provide the correct value of (inverse) `alpha` internally.

in the case of `index_pole_axis`, this is the same for both the forward map and inverse map, so you're right that I can just use the member variables instead, I will fix this!

Let me know if I misunderstood your question :+1:

File: src/Domain/CoordinateMaps/EquatorialCompression.cpp [hunk 8]
```diff
@@ -67,41 +77,47 @@ EquatorialCompression::angular_distortion_jacobian(
...
-  get<1, 2>(jacobian_matrix) = y * z * lambda2;
-  get<2, 2>(jacobian_matrix) = inverse_alpha * (z * z * lambda2 + lambda3);
+  for (size_t i = 0; i < 3; i++) {
+    for (size_t j = 0; j < 3; j++) {
+      const size_t k = (index_pole_axis_ + j) % 3;
+      jacobian_matrix.get(i, k) =
+          gsl::at(coords,i) * gsl::at(coords,k) * (pow<2>(inverse_alpha) - 1.0);
+      if (j == 0) {
+        jacobian_matrix.get(i, k) *=
+            (square(gsl::at(coords,index_pole_axis_)) - square(radius));
```
Comment (knelli2) [inline src/Domain/CoordinateMaps/EquatorialCompression.cpp]: I missed this before, but why do you have both `square` and `pow<2>` here (and below)? I'd say pick one and use that throughout. I think SpECTRE convention is to use `square` and `cube` from `ConstantExpressions.hpp`.

File: tests/Unit/Domain/CoordinateMaps/Test_EquatorialCompression.cpp [hunk 9]
```diff
@@ -46,36 +47,45 @@ void test_radius() {
...
+    // Then, if the argument provided to the EquatorialCompression
+    // map is, for example, 4, points on the undistorted sphere
+    // that have an angle tan \theta = sqrt(x^2 + y^2)/z are
+    // mapped to points that have an angle
+    //  tan \theta' = 4*sqrt(x^2+y^2)/z
+    //              = sqrt(x'^2 + y'^2)/z'.
+
+    const auto& planar_distance =
+        sqrt(pow<2>(input_point[(index_polar_axis + 1) % 3]) +
+             pow<2>(input_point[(index_polar_axis + 2) % 3]));
```
Comment (knelli2) [inline tests/Unit/Domain/CoordinateMaps/Test_EquatorialCompression.cpp]: Clang-tidy still isn't liking this indexing. It suggests `gsl::at` as well.  Here and below too, lines 80-87.

Same comment above about `square` vs `pow<2>`.

**Review by knelli2 — CHANGES_REQUESTED** [review]
Just a couple more things

File: src/Domain/CoordinateMaps/EquatorialCompression.hpp (hunk 6 above)
Comment (knelli2) [reply src/Domain/CoordinateMaps/EquatorialCompression.hpp]: Ah yes I didn't see that the arguments were slightly different, thank you for pointing that out!

**Review by nilsvu — DISMISSED** [review]
Fixup LGTM, though please run clang-format over the files while squashing (an easy way to do this is `git-clang-format upstream/develop`).

**Review by knelli2 — APPROVED** [review]

**Review by nilsvu — APPROVED** [review]
